### PR TITLE
enrich-kinesis: add retry mechanism for the checkpointing (close #591)

### LINKED ISSUE #591: enrich-kinesis: add retry mechanism for the checkpointing
We've already seen the following uncaught exception:
```
java.util.concurrent.TimeoutException: null
    at java.util.concurrent.CompletableFuture.timedGet
    at java.util.concurrent.CompletableFuture.get
    at software.amazon.kinesis.common.FutureUtils.resolveOrCancelFuture(FutureUtils.java:28)
    at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseRefresher.updateLease(DynamoDBLeaseRefresher.java:681)
    at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseRenewer.updateLease(DynamoDBLeaseRenewer.java:314)
    at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseCoordinator.updateLease(DynamoDBLeaseCoordinator.java:353)
    at software.amazon.kinesis.checkpoint.dynamodb.DynamoDBCheckpointer.setCheckpoint(DynamoDBCheckpointer.java:140)
    at software.amazon.kinesis.checkpoint.dynamodb.DynamoDBCheckpointer.setCheckpoint(DynamoDBCheckpointer.java:60)
    at software.amazon.kinesis.checkpoint.ShardRecordProcessorCheckpointer.advancePosition(ShardRecordProcessorCheckpointer.java:291)
    at software.amazon.kinesis.checkpoint.ShardRecordProcessorCheckpointer.checkpoint(ShardRecordProcessorCheckpointer.java:66)
    at com.snowplowanalytics.snowplow.enrich.kinesis.KinesisRun$.$anonfun$checkpoint$5(KinesisRun.scala:76)
    at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
    at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:104)
    at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:463)
    at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:484)
    at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:422)
    at cats.effect.internals.IOShift$Tick.run(IOShift.scala:36)
    at java.util.concurrent.ThreadPoolExecutor.runWorker
    at java.util.concurrent.ThreadPoolExecutor$Worker.run
    at java.lang.Thread.run

software.amazon.kinesis.leases.exceptions.DependencyException: java.util.concurrent.TimeoutException
    at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseRefresher.convertAndRethrowExceptions(DynamoDBLeaseRefresher.java:756)
    at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseRefresher.updateLease(DynamoDBLeaseRefresher.java:692)
    at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseRenewer.updateLease(DynamoDBLeaseRenewer.java:314)
    at software.amazon.kinesis.leases.dynamodb.DynamoDBLeaseCoordinator.updateLease(DynamoDBLeaseCoordinator.java:353)
    at software.amazon.kinesis.checkpoint.dynamodb.DynamoDBCheckpointer.setCheckpoint(DynamoDBCheckpointer.java:140)
    at software.amazon.kinesis.checkpoint.dynamodb.DynamoDBCheckpointer.setCheckpoint(DynamoDBCheckpointer.java:60)
    at software.amazon.kinesis.checkpoint.ShardRecordProcessorCheckpointer.advancePosition(ShardRecordProcessorCheckpointer.java:291)
    at software.amazon.kinesis.checkpoint.ShardRecordProcessorCheckpointer.checkpoint(ShardRecordProcessorCheckpointer.java:66)
    at com.snowplowanalytics.snowplow.enrich.kinesis.KinesisRun$.$anonfun$checkpoint$5(KinesisRun.scala:76)
    at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
    at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:104)
    at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:463)
    at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:484)
    at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:422)
    at cats.effect.internals.IOShift$Tick.run(IOShift.scala:36)
    at java.util.concurrent.ThreadPoolExecutor.runWorker
    at java.util.concurrent.ThreadPoolExecutor$Worker.run
    at java.lang.Thread.run

software.amazon.kinesis.exceptions.KinesisClientLibDependencyException: Unable to save checkpoint for shardId shardId-000000000000
    at software.amazon.kinesis.checkpoint.dynamodb.DynamoDBCheckpointer.setCheckpoint(DynamoDBCheckpointer.java:71)
    at software.amazon.kinesis.checkpoint.ShardRecordProcessorCheckpointer.advancePosition(ShardRecordProcessorCheckpointer.java:291)
    at software.amazon.kinesis.checkpoint.ShardRecordProcessorCheckpointer.checkpoint(ShardRecordProcessorCheckpointer.java:66)
    at com.snowplowanalytics.snowplow.enrich.kinesis.KinesisRun$.$anonfun$checkpoint$5(KinesisRun.scala:76)
    at main$ @ com.snowplowanalytics.snowplow.enrich.kinesis.Main$.cats$effect$internals$IOAppCompanionPlatform$WithContext$$super$main(Main.scala:21)
```

### PR DESCRIPTION
**Pull request by benjben**
I know that we like everything to be configurable, but I wonder if this is worth it for [these ones](https://github.com/snowplow/enrich/pull/592/files#diff-a4a13945ab9a250994b2e255a44ad844eb06dd339220a02d37ff65a46a29d3f7R91-R93), as this is only for checkpointing failures.

### CODE REVIEW

**Comment by istreeter** [comment]
I am mildly (not strongly) opinionated about where to put retry settings.  But my preference is to put it in `application.conf`.

More than anything... it would be great to have a team policy about when it is appropriate to put things in `application.conf` and when it is appropriate to use hard-coded values.  By having a policy, it means we don't need to ask ourselves this question every time!!  Because it's a hard question.

The reasons I like `application.conf` are:
1. All defaults are in one place.  If I need to look up configuration setting then I know where to look.
2. Just in case we need to change it in production, then there is a way.
3. Can't think of any other reason right now, but a list with 3 items is more convincing.

This PR is good, I won't block merging it.  Grab another person's opinion first though.

File: modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/KinesisRun.scala [hunk 1]
```diff
@@ -61,17 +71,34 @@ object KinesisRun {
...
+              if (record.sequenceNumber > biggest.sequenceNumber)
+                record
+              else
+                biggest
+            } :: acc
+        }
+        .parTraverse_(record => Sync[F].delay(record.checkpointer.checkpoint))
+    )
+
+  private def withRetries[F[_]: Sync: Timer, A](action: F[A]): F[A] = {
```
[the remark this answers is not in the document]
Comment (benjben) [reply modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/KinesisRun.scala:90]: I tried to figure out if PubSub checkpointing can fail and should be retried.

We call [record.ack](https://github.com/snowplow/enrich/blob/master/modules/pubsub/src/main/scala/com/snowplowanalytics/snowplow/enrich/pubsub/Main.scala#L79).
`ack` is defined as [such](https://github.com/permutive/fs2-google-pubsub/blob/master/fs2-google-pubsub-grpc/src/main/scala/com/permutive/pubsub/consumer/grpc/internal/Model.scala#L6).
And [this](https://github.com/permutive/fs2-google-pubsub/blob/master/fs2-google-pubsub-grpc/src/main/scala/com/permutive/pubsub/consumer/grpc/internal/PubsubSubscriber.scala#L66) is the implementation.
Then I got to [the interface](https://javadoc.io/static/com.google.cloud/google-cloud-pubsub/1.102.0/com/google/cloud/pubsub/v1/AckReplyConsumer.html#ack--) but I [didn't manage to find](https://javadoc.io/static/com.google.cloud/google-cloud-pubsub/1.102.0/com/google/cloud/pubsub/v1/package-summary.html) the actual implementation and I don't know if there is a retry mechanism in the acking. All the examples that I found use `ack()` with no error handling (e.g. [here](https://javadoc.io/static/com.google.cloud/google-cloud-pubsub/1.102.0/com/google/cloud/pubsub/v1/MessageReceiver.html#receiveMessage-com.google.pubsub.v1.PubsubMessage-com.google.cloud.pubsub.v1.AckReplyConsumer-)), so it seems safe to use. But if it's not, then the stream will fail and crash the app, so we'd better play safe and add the retry to PubSub checkpointing as well.

/cc @istreeter

[the remark this answers is not in the document]
Comment (istreeter) [reply modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/KinesisRun.scala:90]: [This is the implementation](https://github.com/googleapis/java-pubsub/blob/ea5a48d16010d3e12af4b9a6281feb44a5f46b7f/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java#L443-L453) of that interface.

I also noticed [this interesting bit of code](https://github.com/googleapis/java-pubsub/blob/ea5a48d16010d3e12af4b9a6281feb44a5f46b7f/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java#L137-L146).  It looks like if it fails to `ack` the message then it immediately `nack`s the message instead.  Does that change at all your decision to retry acking?  Does this need more research, or is it more trouble than it's worth?

[the remark this answers is not in the document]
Comment (istreeter) [reply modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/KinesisRun.scala:90]: Oh.... ignore that comment.  I mis-read the code.  It only nacks if it fails to process the original message, not if it fails to ack.

[the remark this answers is not in the document]
Comment (istreeter) [reply modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/KinesisRun.scala:90]: I think [this is the class that processes acks](https://github.com/googleapis/java-pubsub/blob/ea5a48d16010d3e12af4b9a6281feb44a5f46b7f/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java#L53).  And on a quick scan [it looks like it handles retries](https://github.com/googleapis/java-pubsub/blob/ea5a48d16010d3e12af4b9a6281feb44a5f46b7f/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java#L558).

[the remark this answers is not in the document]
Comment (benjben) [reply modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/KinesisRun.scala:90]: Thanks @istreeter ! From what I can see in the code there is good error handling everywhere, so even if the ack fails, there won't be an exception thrown and we won't be able to retry, so I'll add the retry only for Kinesis.

[the remark this answers is not in the document]
Comment (benjben) [reply modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/KinesisRun.scala:90]: I hadn't seen your previous comment. Great!

**Review by istreeter — APPROVED** [review]

**Review by oguzhanunlu — APPROVED** [review]
Thanks Ben!

**Comment by oguzhanunlu** [comment]
just noticed the scalafmt error, could you fix that @benjben ?

**Pull request merged**